File: samples/android/AdUnitsSample/src/main/java/com/facebook/samples/AdUnitsSample/fragments/NativeAdTemplateFragment.java
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 * All rights reserved.
 *
 * This source code is licensed under the license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.samples.AdUnitsSample.fragments;

import static android.view.ViewGroup.LayoutParams.MATCH_PARENT;

import android.app.Activity;
import android.content.res.Resources;
import android.graphics.Color;
import android.os.Bundle;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.Button;
import android.widget.SeekBar;
import android.widget.Spinner;
import android.widget.TextView;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import com.facebook.ads.Ad;
import com.facebook.ads.AdError;
import com.facebook.ads.NativeAd;
import com.facebook.ads.NativeAdListener;
import com.facebook.ads.NativeAdView;
import com.facebook.ads.NativeAdViewAttributes;
import com.facebook.samples.AdUnitsSample.R;

public class NativeAdTemplateFragment extends Fragment implements NativeAdListener {

  private static final String TAG = NativeAdTemplateFragment.class.getSimpleName();

  private static final int COLOR_LIGHT_GRAY = 0xff90949c;
  private static final int COLOR_DARK_GRAY = 0xff4e5665;
  private static final int COLOR_CTA_BLUE_BG = 0xff4080ff;

  private static final int MIN_HEIGHT_DP = 200;
  private static final int MAX_HEIGHT_DP = 500;
  private static final int DEFAULT_HEIGHT_DP = 350;
  private static final int DEFAULT_PROGRESS_DP = 50;

  private @Nullable NativeAd mNativeAd;
  private int mLayoutHeightDp = DEFAULT_HEIGHT_DP;

  private int mAdBackgroundColor, mTitleColor, mCtaTextColor, mContentColor, mCtaBgColor;

  private TextView mStatusText;
  private ViewGroup mNativeAdContainer;
  private Spinner mBackgroundColorSpinner;
  private Button mShowCodeButton, mReloadButton;
  private SeekBar mSeekBar;
  private View mAdView;

  @Override
  public View onCreateView(
      LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {

    View view = inflater.inflate(R.layout.fragment_native_ad_template, container, false);

    mStatusText = view.findViewById(R.id.status);
    mNativeAdContainer = view.findViewById(R.id.templateContainer);
    mShowCodeButton = view.findViewById(R.id.showCodeButton);
    mReloadButton = view.findViewById(R.id.reloadAdButton);
    mBackgroundColorSpinner = view.findViewById(R.id.backgroundColorSpinner);
    mSeekBar = view.findViewById(R.id.seekBar);

    setUpLayoutBuilders();
    setUpButtons();

    createAndLoadNativeAd();

    return view;
  }

  @Override
  public void onAdLoaded(Ad ad) {
    if (mNativeAd == null || mNativeAd != ad) {
      // Race condition, load() called again before last ad was displayed
      return;
    }

    mStatusText.setText(R.string.ad_loaded);
    reloadAdContainer();
  }

  @Override
  public void onError(Ad ad, AdError error) {
    String msg = getResources().getString(R.string.ad_load_failed, error.getErrorMessage());
    mStatusText.setText(msg);
  }

  @Override
  public void onAdClicked(Ad ad) {
    mStatusText.setText(R.string.ad_clicked);
  }

  @Override
  public void onLoggingImpression(Ad ad) {
    Log.d(TAG, "onLoggingImpression");
  }

  @Override
  public void onMediaDownloaded(Ad ad) {
    Log.d(TAG, "onMediaDownloaded");
  }

  @Override
  public void onDestroy() {
    mNativeAd = null;
    super.onDestroy();
  }

  private void createAndLoadNativeAd() {
    // Create a native ad request with a unique placement ID
    // (generate your own on the Facebook app settings).
    // Use different ID for each ad placement in your app.
    mNativeAd = new NativeAd(getActivity(), "YOUR_PLACEMENT_ID");

    // Initiate a request to load an ad.
    mNativeAd.loadAd(mNativeAd.buildLoadAdConfig().withAdListener(this).build());

    mStatusText.setText(R.string.ad_loading);
  }

  private void reloadAdContainer() {
    Activity activity = getActivity();
    if (activity != null
        && mNativeAd != null
        && mNativeAd.isAdLoaded()
        && !mNativeAd.isAdInvalidated()) {
      mNativeAdContainer.removeAllViews();

      // Create a NativeAdViewAttributes object and set the attributes
      NativeAdViewAttributes attributes =
          new NativeAdViewAttributes()
              .setBackgroundColor(mAdBackgroundColor)
              .setTitleTextColor(mTitleColor)
              .setDescriptionTextColor(mContentColor)
              .setButtonBorderColor(mCtaTextColor)
              .setButtonTextColor(mCtaTextColor)
              .setButtonColor(mCtaBgColor);

      // Use NativeAdView.render to generate the ad View
      mAdView = NativeAdView.render(activity, mNativeAd, attributes);

      mNativeAdContainer.addView(mAdView, new ViewGroup.LayoutParams(MATCH_PARENT, 0));
      updateAdViewParams();

      mShowCodeButton.setText(R.string.show_code);
    }
  }

  private void setUpLayoutBuilders() {
    ArrayAdapter<CharSequence> backgroundColorSpinnerAdapter =
        ArrayAdapter.createFromResource(
            getActivity(), R.array.background_color_array, android.R.layout.simple_spinner_item);
    backgroundColorSpinnerAdapter.setDropDownViewResource(
        android.R.layout.simple_spinner_dropdown_item);
    mBackgroundColorSpinner.setAdapter(backgroundColorSpinnerAdapter);

    mBackgroundColorSpinner.setOnItemSelectedListener(
        new android.widget.AdapterView.OnItemSelectedListener() {
          @Override
          public void onItemSelected(AdapterView<?> arg0, View view, int position, long id) {
            int item = mBackgroundColorSpinner.getSelectedItemPosition();
            switch (item) {
              case 0:
                mAdBackgroundColor = Color.WHITE;
                mTitleColor = COLOR_DARK_GRAY;
                mCtaTextColor = COLOR_CTA_BLUE_BG;
                mContentColor = COLOR_LIGHT_GRAY;
                mCtaBgColor = Color.WHITE;
                break;
              case 1:
                mAdBackgroundColor = Color.BLACK;
                mTitleColor = Color.WHITE;
                mContentColor = Color.LTGRAY;
                mCtaTextColor = Color.BLACK;
                mCtaBgColor = Color.WHITE;
                break;
            }
            reloadAdContainer();
          }

          @Override
          public void onNothingSelected(AdapterView<?> parent) {}
        });

    mSeekBar.setProgress(DEFAULT_PROGRESS_DP);
    mSeekBar.setOnSeekBarChangeListener(
        new SeekBar.OnSeekBarChangeListener() {
          @Override
          public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {
            mLayoutHeightDp = progress * ((MAX_HEIGHT_DP - MIN_HEIGHT_DP) / 100) + MIN_HEIGHT_DP;
            updateAdViewParams();
          }

          @Override
          public void onStartTrackingTouch(SeekBar seekBar) {}

          @Override
          public void onStopTrackingTouch(SeekBar seekBar) {}
        });
  }

  private void updateAdViewParams() {
    if (mAdView == null) {
      return;
    }
    ViewGroup.LayoutParams params = mAdView.getLayoutParams();
    params.height = (int) (Resources.getSystem().getDisplayMetrics().density * mLayoutHeightDp);
    mAdView.setLayoutParams(params);
    mAdView.requestLayout();
  }

  private void setUpButtons() {
    mShowCodeButton.setOnClickListener(
        new View.OnClickListener() {
          @Override
          public void onClick(View v) {
            if (mShowCodeButton.getText() == getResources().getString(R.string.show_ad)) {
              reloadAdContainer();
            } else {
              showCodeInAdContainer();
            }
          }
        });

    mReloadButton.setOnClickListener(
        new View.OnClickListener() {
          @Override
          public void onClick(View v) {
            createAndLoadNativeAd();
          }
        });
  }

  private void showCodeInAdContainer() {
    String[] lines = getResources().getStringArray(R.array.code_snippet_mediumrect_template);
    StringBuilder codeSnippet = new StringBuilder();
    for (String line : lines) {
      codeSnippet.append(line).append("\r\n");
    }
    mNativeAdContainer.removeAllViews();
    TextView code = new TextView(getActivity());
    code.setText(codeSnippet);
    code.setBackgroundColor(Color.WHITE);
    code.setTextColor(Color.BLACK);
    mNativeAdContainer.addView(code, 0);

    mShowCodeButton.setText(R.string.show_ad);
  }
}

File: samples/android/KotlinAdUnitsSample/src/main/java/com/facebook/samples/adunitssamplekotlin/fragments/NativeAdTemplateFragment.kt
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 * All rights reserved.
 *
 * This source code is licensed under the license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.samples.adunitssamplekotlin.fragments

import android.content.Context
import android.content.res.Resources
import android.graphics.Color
import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.*
import androidx.fragment.app.Fragment
import com.facebook.ads.*
import com.facebook.samples.adunitssamplekotlin.R

class NativeAdTemplateFragment : Fragment(), NativeAdListener {

  private var nativeAd: NativeAd? = null
  private var layoutHeightDp = DEFAULT_HEIGHT_DP

  private var adBackgroundColor: Int = 0
  private var titleColor: Int = 0
  private var ctaTextColor: Int = 0
  private var contentColor: Int = 0
  private var ctaBgColor: Int = 0

  private var statusText: TextView? = null
  private var nativeAdContainer: ViewGroup? = null
  private var backgroundColorSpinner: Spinner? = null
  private var showCodeButton: Button? = null
  private var reloadButton: Button? = null
  private var seekBar: SeekBar? = null
  private var adView: View? = null

  override fun onCreateView(
      inflater: LayoutInflater,
      container: ViewGroup?,
      savedInstanceState: Bundle?
  ): View? {

    val view = inflater.inflate(R.layout.fragment_native_ad_template, container, false)

    statusText = view.findViewById(R.id.status)
    nativeAdContainer = view.findViewById(R.id.templateContainer)
    showCodeButton = view.findViewById(R.id.showCodeButton)
    reloadButton = view.findViewById(R.id.reloadAdButton)
    backgroundColorSpinner = view.findViewById(R.id.backgroundColorSpinner)
    seekBar = view.findViewById(R.id.seekBar)

    setUpLayoutBuilders(inflater.context)
    setUpButtons()

    createAndLoadNativeAd()

    return view
  }

  override fun onAdLoaded(ad: Ad) {
    if (nativeAd == null || nativeAd !== ad) {
      // Race condition, load() called again before last ad was displayed
      return
    }

    statusText?.setText(R.string.ad_loaded)
    reloadAdContainer()
  }

  override fun onError(ad: Ad, error: AdError) {
    val msg = resources.getString(R.string.ad_load_failed, error.errorMessage)
    statusText?.text = msg
  }

  override fun onAdClicked(ad: Ad) {
    statusText?.setText(R.string.ad_clicked)
  }

  override fun onLoggingImpression(ad: Ad) {
    Log.d(TAG, "onLoggingImpression")
  }

  override fun onMediaDownloaded(ad: Ad) {
    Log.d(TAG, "onMediaDownloaded")
  }

  override fun onDestroy() {
    nativeAd = null
    super.onDestroy()
  }

  private fun createAndLoadNativeAd() {
    // Create a native ad request with a unique placement ID
    // (generate your own on the Facebook app settings).
    // Use different ID for each ad placement in your app.
    nativeAd = NativeAd(activity, "YOUR_PLACEMENT_ID")

    // Initiate a request to load an ad.
    nativeAd!!.loadAd(
        nativeAd!!
            .buildLoadAdConfig()
            // Set a listener to get notified when the ad was loaded.
            .withAdListener(this)
            .build())

    statusText?.setText(R.string.ad_loading)
  }

  private fun reloadAdContainer() {
    val activity = activity
    if (activity != null && nativeAd != null && nativeAd!!.isAdLoaded) {
      nativeAdContainer!!.removeAllViews()

      // Create a NativeAdViewAttributes object and set the attributes
      val attributes =
          NativeAdViewAttributes(context)
              .setBackgroundColor(adBackgroundColor)
              .setTitleTextColor(titleColor)
              .setDescriptionTextColor(contentColor)
              .setButtonBorderColor(ctaTextColor)
              .setButtonTextColor(ctaTextColor)
              .setButtonColor(ctaBgColor)

      // Use NativeAdView.render to generate the ad View
      adView = NativeAdView.render(activity, nativeAd!!, attributes)

      nativeAdContainer!!.addView(
          adView, ViewGroup.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, 0))
      updateAdViewParams()

      showCodeButton!!.setText(R.string.show_code)
    }
  }

  private fun setUpLayoutBuilders(context: Context) {
    val backgroundColorSpinnerAdapter =
        ArrayAdapter.createFromResource(
            context, R.array.background_color_array, android.R.layout.simple_spinner_item)
    backgroundColorSpinnerAdapter.setDropDownViewResource(
        android.R.layout.simple_spinner_dropdown_item)
    backgroundColorSpinner!!.adapter = backgroundColorSpinnerAdapter

    backgroundColorSpinner!!.onItemSelectedListener =
        object : AdapterView.OnItemSelectedListener {
          override fun onItemSelected(arg0: AdapterView<*>, view: View, position: Int, id: Long) {
            when (backgroundColorSpinner!!.selectedItemPosition) {
              0 -> {
                adBackgroundColor = Color.WHITE
                titleColor = COLOR_DARK_GRAY
                ctaTextColor = COLOR_CTA_BLUE_BG
                contentColor = COLOR_LIGHT_GRAY
                ctaBgColor = Color.WHITE
              }
              1 -> {
                adBackgroundColor = Color.BLACK
                titleColor = Color.WHITE
                contentColor = Color.LTGRAY
                ctaTextColor = Color.BLACK
                ctaBgColor = Color.WHITE
              }
            }
            reloadAdContainer()
          }

          override fun onNothingSelected(parent: AdapterView<*>) = Unit
        }

    seekBar!!.progress = DEFAULT_PROGRESS_DP
    seekBar!!.setOnSeekBarChangeListener(
        object : SeekBar.OnSeekBarChangeListener {
          override fun onProgressChanged(seekBar: SeekBar, progress: Int, fromUser: Boolean) {
            layoutHeightDp = progress * ((MAX_HEIGHT_DP - MIN_HEIGHT_DP) / 100) + MIN_HEIGHT_DP
            updateAdViewParams()
          }

          override fun onStartTrackingTouch(seekBar: SeekBar) = Unit

          override fun onStopTrackingTouch(seekBar: SeekBar) = Unit
        })
  }

  private fun updateAdViewParams() {
    if (adView == null) {
      return
    }
    val params = adView!!.layoutParams
    params.height = (Resources.getSystem().displayMetrics.density * layoutHeightDp).toInt()
    adView!!.layoutParams = params
    adView!!.requestLayout()
  }

  private fun setUpButtons() {
    showCodeButton!!.setOnClickListener {
      if (showCodeButton!!.text === resources.getString(R.string.show_ad)) {
        reloadAdContainer()
      } else {
        showCodeInAdContainer()
      }
    }

    reloadButton!!.setOnClickListener { createAndLoadNativeAd() }
  }

  private fun showCodeInAdContainer() {
    val lines = resources.getStringArray(R.array.code_snippet_mediumrect_template)
    val codeSnippet = StringBuilder()
    for (line in lines) {
      codeSnippet.append(line).append("\r\n")
    }
    nativeAdContainer!!.removeAllViews()
    val code = TextView(activity)
    code.text = codeSnippet
    code.setBackgroundColor(Color.WHITE)
    code.setTextColor(Color.BLACK)
    nativeAdContainer!!.addView(code, 0)

    showCodeButton!!.setText(R.string.show_ad)
  }

  companion object {

    private val TAG = NativeAdTemplateFragment::class.java.simpleName

    private const val COLOR_LIGHT_GRAY = -0x6f6b64
    private const val COLOR_DARK_GRAY = -0xb1a99b
    private const val COLOR_CTA_BLUE_BG = -0xbf7f01

    private const val MIN_HEIGHT_DP = 200
    private const val MAX_HEIGHT_DP = 500
    private const val DEFAULT_HEIGHT_DP = 350
    private const val DEFAULT_PROGRESS_DP = 50
  }
}
